set.mm

```diff
@@ -732651,6 +732663,410 @@ The modulo (remainder) operation - extension
+  used.  As a special case, the idiom for the preimage of a function value at
+  ` X ` under a function ` F ` is ` ( ``' F " { ( F `` X ) } ) ` (according to
+  Wikipedia, the preimage of a singleton is also called a "fiber").
+
+  We use the label fragment "preima" (as in ~ mptpreima ) for theorems about
+  preimages (sometimes, also "imacnv" is used as in ~ fvimacnvi ), and
+  "preimafv" (as in ~ preimafvn0 ) for theorems about preimages of a function
+  value.
+
+  In this section, ` P = { z | E. x e. A z = ( ``' F " { ( F `` x ) } ) } `
+  will be the set of all preimages of function values of a function ` F `,
+  that means ` S e. P ` is a preimage of a function value (see, for example,
+  ~ elsetpreimafv ): ` S = ( ``' F " { ( F `` x ) } ) `.
+
+  With the help of such a set, it is shown that every function ` F : A --> B `
+  can be decomposed into a surjective and an injective function (see
+  ~ fundcmpsurinj ) by constructing a surjective function ` g : A -onto-> P `
+  and an injective function ` h : P -1-1-> B ` so that ` F = ( h o. g ) ` ( see
+  ~ fundcmpsurinjpreimafv ).  See also Wikipedia ("Surjective function",
+  17-Mar-2024, ~ https://en.wikipedia.org/wiki/Surjective_function (section
+  "Composition and decomposition").  This is different from the decomposition
+  of ` F ` into the surjective function ` g : A -onto-> ( F " A ) ` (with
+  ` ( g `` x ) = ( F `` x ) ` for ` x e. A `) and the injective function
+  ` h = ( _I |`` ( F " A ) ) `,  ( see ~ fundcmpsurinjimaid ), see also
+  Wikipedia ("Bijection, injection and surjection", 17-Mar-2024,
+  ~ https://en.wikipedia.org/wiki/Bijection,_injection_and_surjection (section
+  "Properties").
+
+$)
+
+  $( The preimage of a function value at ` X ` contains ` X ` .  (Contributed
+     by AV, 7-Mar-2024.) $)
+  preimafvsnel $p |- ( ( F Fn A /\ X e. A )
+                       -> X e. ( `' F " { ( F ` X ) } ) ) $=
+    ( wfn wcel wa ccnv cfv csn cima wceq simpr eqidd fniniseg adantr mpbir2and
+    wb ) BADZCAEZFZCBGCBHZIJEZSUAUAKZRSLTUAMRUBSUCFQSAUACBNOP $.
+
+  $( The preimage of a function value is not empty.  (Contributed by AV,
+     7-Mar-2024.) $)
+  preimafvn0 $p |- ( ( F Fn A /\ X e. A )
+                     -> ( `' F " { ( F ` X ) } ) =/= (/) ) $=
+    ( wfn wcel wa ccnv cfv csn cima preimafvsnel ne0d ) BADCAEFBGCBHIJCABCKL $.
+
+  ${
+    $d F x y $.  $d S x y $.  $d X x y $.
+    $( The union of the image of a subset ` S ` of the domain of a Function
+       with elements having the same function value is the function value at
+       one of the elements of ` S ` .  (Contributed by AV, 5-Mar-2024.) $)
+    uniimafveqt $p |- ( ( F : A --> B /\ S C_ A /\ X e. S )
+                       -> ( A. x e. S ( F ` x ) = ( F ` X )
+                          -> U. ( F " S ) = ( F ` X ) ) ) $=
+      ( vy wf wss wcel w3a cv cfv wceq wral cima cuni wa ciun wfun 3ad2ant1 syl
+      ffun adantr funiunfv simp3 fveqeq2 cbvralv biimpi fveq2 iuneqconst syl2an
```

Comment:
you can use [cbvralvw](https://us.metamath.org/mpeuni/cbvralvw.html) instead of [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) here, which will let you avoid ax-13 without adding any extra dv conditions:

<pre>
  ${
    $d F x y $.  $d S x y $.  $d X x y $.
    $( The union of the image of a subset ` S ` of the domain of a Function
       with elements having the same function value is the function value at
       one of the elements of ` S ` .  (Contributed by AV, 5-Mar-2024.) $)
    uniimafveqt $p |- ( ( F : A --> B /\ S C_ A /\ X e. S )
                       -> ( A. x e. S ( F ` x ) = ( F ` X )
                          -> U. ( F " S ) = ( F ` X ) ) ) $=
     ( vy wf wss wcel w3a cv cfv wceq wral cima cuni wa ciun wfun 3ad2ant1 syl
      ffun adantr funiunfv simp3 cbvralvw biimpi fveq2 iuneqconst syl2an eqtr3d
      fveqeq2 ex ) BCEHZDBIZFDJZKZALZEMFEMZNZADOZEDPQZUTNURVBRZGDGLZEMZSZVCUTVD
      ETZVGVCNURVHVBUOUPVHUQBCEUCUAUDGDEUEUBURUQVFUTNZGDOZVGUTNVBUOUPUQUFVBVJVA
      VIAGDUSVEUTEUMUGUHGDVFUTFVEFEUIUJUKULUN $.
$}
</pre>

Actually, I think [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) should be discouraged, and we should invite people to use [cbvralvw](https://us.metamath.org/mpeuni/cbvralvw.html) instead.

Reply:
I like the idea of making this a discouragement tag if we expect people to follow this practice generally.

Reply:
Done - I do not know if this was discussed already before (maybe in Google group: https://groups.google.com/g/metamath/c/OB2_9sYgDfA.?), but I would also endorse to tag cbvralv and cbvrexv (are there more?) as "discouraged".

Reply:
The discouraged tag is very prohibitive.  The kind of problem given here appears very often: alrimi vs alrimih, nfsb vs nfsbv and so on, leading a very long (and long standing) list.  The idea behind any of them is that more distinct variables allow for a smaller axiom footprint.  Or: more axioms can provide a more general theorem, which is also a valuable goal.
Here we have the situation that the dv conditions already stated allow for the narrower theorem. What you need is a kind of pick list in your proof designer, that displays close variants of a theorem, and where you easily find the best choice.  Without such an assistant all you can do is improve on documentation.  It would be helpful if you see an overview of variants in a particular location, maybe a doc only web page like "mathbox", but dedicated to cbvral* versions.

Reply:
@GinoGiotto has spent a lot of effort to remove references to ax-13. 
If the [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) variant is not discouraged, then it's like trying to fill a leaky bucket: references to ax-13 will always come back. So I'd also be in favor of marking [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) as discouraged in profit of [cbvralvw](https://us.metamath.org/mpeuni/cbvralvw.html), instead of checking every new incoming PR.

Said differently, if both theorems are applicable, which would be a reason to use [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) over [cbvralvw](https://us.metamath.org/mpeuni/cbvralvw.html)?

Reply:
`if both theorems are applicable, which would be a reason to use [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) over [cbvralvw](https://us.metamath.org/mpeuni/cbvralvw.html)`
There is no reason to use cbvralv without any advantage.  I looked at a random subset of its applications and found no instance, where it was not replaceable.  So maybe, cbvralv has no other value than stating that one dv condition in theory is unnecessary (but in practice no one cares).
But this is not an isolated problem.  In this PR alone we have another candidate cbvrexv, where we can start the discussion over again. And far more need such considerations, not alone in respect to ax-13, but other axioms as well.  So simply discouraging anything but the variant with most dv restrictions is not a viable option, in my eyes.
As always, you need to look out for optimizations yourself, and within a mathbox you  are free the freedom to just sketch a proof.  Only when you move sth to main, some care is in order.
I myself put a theorem wl-cbvalsbi into my mathbox, with unnecessary ax-10, ax-11 and ax-12 references in its proof, only to improve on it later.
That's the way work is done up to now.
Given that an unnecessary reference to ax-13 is a magnitude greater than to, say, ax-11, what do you think of extending the $j annotation to mark a deliberate use of ax-13? $j need ax-13 $. for example.  And a verifier that checks the simultaneous presence of ax-13 in the axiom list and this annotation.

Reply:
> Done - I do not know if this was discussed already before (maybe in Google group: [groups.google.com/g/metamath/c/OB2_9sYgDfA](https://groups.google.com/g/metamath/c/OB2_9sYgDfA).?)

This hasn't been discussed yet, but looking at the discussion above it seems this topic deserves its own issue.

Given that ax-13 usage has recently reached an all time low, I would like to have some verification system that it will stay low, not only in the near future, but possibily in the years to come. If we don't do anything about it, I can assure you ax-13 usage will rise again. I've been collecting data about ax-13 usage for a while now. The lesson I've derived is: if you don't systematically check for something, undesired trends will show up, even if small or temporary (and they are usually temporary only because there is someone that checks for it afterwards, which should not be taken for granted).

I'm not particularly attached to the discouragement system to check for this metric, it's just the easiest tool it is currently available for us. If someone decides to show up and build a tool just for this problem, I'm not against it, on the contrary. But it has to be automatic and systematic, human review is not enough.

I don't want to give the impression that using ax-13 is prohibited. In the end, axioms are there to be used. I want to avoid the scenario in which ax-13 is used without awareness. So rather than a prohibition issue this is a communication issue. We need to find a way to communicate the user "hey, it seems you're trying to build your proof using ax-13. Did you know that about 97% of theorems in the database don't need it? Perhaps yours doesn't need it too."

Wolf Lammen is right to say that this problem does not only affects [cbvralv](https://us.metamath.org/mpeuni/cbvralv.html) or [cbvrexv](https://us.metamath.org/mpeuni/cbvrexv.html), this affects dozens of theorems, so either we are down to add about a hundred or so of discouragement tags (personally I don't have a problem with that), or we build something else. The worst decision in my opinion is to leave the state of the art as it is.

Reply:
`if you don't systematically check for something, undesired trends will show up, even if small or temporary`
It will take some time, sure, but maybe after a while people will routinely carry out checks of this kind?  There is evidence for this in the history of Metamath, which has seen some evolvement with deprecating patterns.